packages/ckeditor5-engine/tests/controller/datacontroller.js

```diff
@@ -671,6 +671,78 @@ describe( 'DataController', () => {
 			expect( stringifyView( viewDocumentFragment ) ).to.equal( 'f<span class="a">oo</span>' );
 		} );
 
+		// See https://github.com/ckeditor/ckeditor5/issues/8485.
+		it( 'should fire an addMarker event for collapsed markers located at $root element boundary', () => {
+			const root = model.document.getRoot();
+			const spy = sinon.spy();
+
+			data.downcastDispatcher.on( 'addMarker:fooMarkerAtElementStart', spy );
+			data.downcastDispatcher.on( 'addMarker:fooMarkerAtElementEnd', spy );
+
+			setData( model, '<paragraph>foobar</paragraph>' );
+
+			model.change( writer => {
+				writer.addMarker( 'fooMarkerAtElementStart', {
+					range: writer.createRange(
+						writer.createPositionFromPath( root, [ 0 ] ),
+						writer.createPositionFromPath( root, [ 0 ] )
+					),
+					usingOperation: true
+				} );
+
+				writer.addMarker( 'fooMarkerAtElementEnd', {
+					range: writer.createRange(
+						writer.createPositionFromPath( root, [ 1 ] ),
+						writer.createPositionFromPath( root, [ 1 ] )
+					),
+					usingOperation: true
+				} );
+			} );
+
+			data.toView( root );
```

Comment:
Let's check how this looks in the view as in other tests it's done.

Reply:
I rewrote the tests to make them look similar to others. I didn't use the `markerToHighlight` conversion, because it seems that it does not allow creating a collapsed marker. Instead, I'm using now `markerToData` helper.